python/langsmith/anonymizer.py

```diff
@@ -0,0 +1,182 @@
+
+    deep_clone: Optional[bool]
+    """Deep clone the data before replacing."""
+
+
+class StringNodeRule(TypedDict):
+    """Declarative rule used for replacing sensitive data."""
+
+    pattern: re.Pattern
+    """Regex pattern to match."""
+
+    replace: Optional[str]
+    """Replacement value. Defaults to `[redacted]` if not specified."""
+
+
+class RuleNodeProcessor(StringNodeProcessor):
+    """String node processor that uses a list of rules to replace sensitive data."""
+
+    rules: List[StringNodeRule]
+
+    def __init__(self, rules: List[StringNodeRule]):
+        """Initialize the processor with a list of rules."""
+        self.rules = rules
+
+    def mask_nodes(self, nodes: List[StringNode]) -> List[StringNode]:
+        """Mask nodes using the rules."""
+        result = []
+        for item in nodes:
+            new_value = item["value"]
+            for rule in self.rules:
+                new_value = rule["pattern"].sub(
+                    (
+                        rule["replace"]
+                        if isinstance(rule["replace"], str)
+                        else "[redacted]"
+                    ),
+                    new_value,
+                )
+            if new_value != item["value"]:
+                result.append(StringNode(value=new_value, path=item["path"]))
+        return result
+
+
+class CallableNodeProcessor(StringNodeProcessor):
+    """String node processor that uses a callable function to replace sensitive data."""
+
+    func: Union[Callable[[str], str], Callable[[str, List[Union[str, int]]], str]]
+    accepts_path: bool
+
+    def __init__(
+        self,
+        func: Union[Callable[[str], str], Callable[[str, List[Union[str, int]]], str]],
+    ):
+        """Initialize the processor with a callable function."""
+        self.func = func
+        self.accepts_path = len(inspect.signature(func).parameters) == 2
+
+    def mask_nodes(self, nodes: List[StringNode]) -> List[StringNode]:
+        """Mask nodes using the callable function."""
+        retval: List[StringNode] = []
+        for node in nodes:
+            candidate = (
+                self.func(node["value"], node["path"])  # type: ignore[call-arg]
+                if self.accepts_path
+                else self.func(node["value"])  # type: ignore[call-arg]
+            )
+            if candidate != node["value"]:
+                retval.append(StringNode(value=candidate, path=node["path"]))
+        return retval
+
+
+ReplacerType = Union[
+    Callable[[str, List[Union[str, int]]], str],
+    List[StringNodeRule],
+    StringNodeProcessor,
+]
+
+
+def _get_node_processor(replacer: ReplacerType) -> StringNodeProcessor:
+    if isinstance(replacer, list):
+        return RuleNodeProcessor(rules=replacer)
+    elif callable(replacer):
+        return CallableNodeProcessor(func=replacer)
+    else:
+        return replacer
+
+
+def create_anonymizer(
+    replacer: ReplacerType, options: Optional[ReplacerOptions] = None
```

Comment:
it seems more pythonic to have the options as separate keyword args (would also let the user avoid having to decide "deep_clone")